### PR TITLE
msg: use a factory for ease of managing all message types

### PR DESCRIPTION
**Pull request by qicosmos**
the hundreds of switch-case statment has been reduced by a self-register factory.
If you want to add a new message type, just need define a simple micro like other message.

### CODE REVIEW

**Comment by tchaikov** [comment]
@qicosmos 

- did you `git add` the MessageFactory.h?
- could you add a prefix for the first line of your commit message, `msg: ` would suffice.
- could also add a `Fixed: #14202` line right before your `Signed-off-by` line.

**Comment by qicosmos** [comment]
@tchaikov 
hi, i miss the MessageFactory.h, i'll add the file right now and will commit follow your rule.
could i close this pr, and create a new pr?

**Comment by tchaikov** [comment]
@qicosmos you can simply update your local branch, and then `git push -f` to your remote branch, and this pull request will gets updated automatically.

**Comment by mattbenjamin** [comment]
cbodley has a change from over a year ago which adds a MessageFactory, but it is a performance improvement as well as a cleanup;  We should merge the other version, but acknowledge this, maybe.

**Comment by cbodley** [comment]
I'm skeptical that std::map lookups will be faster than what the optimizer does to the switch statement. This is called for every message, so performance is going to be more important than cleanliness. Have you done any benchmarking with this?

I also worked on a MessageFactory, but my strategy was different.. Instead of being global, each daemon had its own factory. Those were still switch statements, but they only included the message types that daemon was interested in. This approach also supported another optimization, which was to combine some internal state with the message in a single allocation; the main example being the osd factory creating an OpRequest instead of MOSDOp.

That said, if you do have performance comparisons that show an improvement, I'd be fine starting with this. Consider trying with std::unordered_map instead of std::map, as well.

**Comment by mattbenjamin** [comment]
Seems like a red herring.  Other than the mixed-in message allocation change, what was blocking merge of the original MessageFactory?

**Comment by cbodley** [comment]
@mattbenjamin It was against a firefly branch. Lots of message types have changed, so the rebase would be messy and prone to missing things. When I looked at it a few months ago, I decided it would be best to redo the work from scratch. But I didn't see it as a priority, especially now that the OpRequest trick is no longer valid.

**Comment by mattbenjamin** [comment]
would be good not to lose the work...somehow

**Comment by gregsfortytwo** [comment]
I presume via the self-registration we'd only include messages relevant to each daemon, assuming it stops including each one in Message.cc

And I doubt that choosing the right decoding functions efficiently is an important part of the cost of message decode+delivery in Ceph, but it would be good to check with some basic benchmarks. I don't think we have anything better than running ceph_test_rados against a MemoryStore-based OSD, though. :/

**Comment by qicosmos** [comment]
hi
I will do a benkmark, in theroy switch-case statement is a little faster than unordered_map, but i think the performance is very close.
Iill give the comparisons latter.
I think the biggest benefit is writing the shortest code, no need to modify the original code, and add new message is very simple and safe. Don't worry the performance, it will never be bottlenek.

**Comment by qicosmos** [comment]
Now that you mention performance, i have some advice:
1.some container pass by value instead of reference.
2.list is used ervery where, however vector is more faster than list in push_back.
3.some temporary variables shuould be replaced by rvalue and move semantics.
4.some functions should be improved by constexpr.

**Comment by qicosmos** [comment]
I have make a benkmak, loop 14million times, here is the result:
switch-case: 0.66s
messagefactory:1.1s
Not surprisin switch-case is 40% faster, but it just cost more 0.4s loop 14million times, so the messagefactory make no difference to performance.

File: src/msg/MessageFactory.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        register_t(int key, Args... args)
+        {
+            MessageFactory::get().map_.emplace(key, [&] { return new T(args...); });
+        }
+        inline static Message* create() { return new T; }
+    };
+
+    inline Message* produce(int key)
+    {
+        if (map_.find(key) == map_.end())
```
Comment (cbodley) [inline src/msg/MessageFactory.h]: if you store the iterator returned by `find()`, you can avoid doing another binary search below with `map_[key]`

**Comment by cbodley** [comment]
@qicosmos all good advice! i'd love to see more of those changes as well

thanks for running the benchmark. did you try with unordered_map?

could you also remove all of the message #includes that are no longer needed in Message.cc? @gregsfortytwo pointed out that, without those, each daemon will only register the messages it actually includes

**Comment by qicosmos** [comment]
@cbodley 
Thanks for you advice, using iterator is better.
Benkmark is based on map not unordered_map, i also test unordered_map, but when the key is integer, map is faster than unordered_map.
I will remove redundant message #includes in Message.cc later.

**Comment by gregsfortytwo** [comment]
So I think we probably want to merge this once the Jewel freeze is done. But that probably won't happen until all the includes get cleaned up (which will also improve the map's performance relative to switch statement!), and it's not a big deal but if it were rebased on top of current master instead of having them be merged in that'd make it easier for people doing code archaeology in the future. :)

**Comment by qicosmos** [comment]
@gregsfortytwo 
I hope the code will be merged on the Jewel freeze, i'll clean up the includes in 3 days.

**Comment by qicosmos** [comment]
@gregsfortytwo
hi，i have finish the work.
To use msg factory need two step, first include the concrete message header file, second add micro at the end of concrete message. 
so the includes no need to remove, if you forget adding the include, the message won't to be register.

**Comment by mattbenjamin** [comment]
This PR still assumes a global MessageFactory at compile time as I read the code.  I would like this change to absorb the concept of contextual (e.g., per-daemon) message factories, so we get message compression to just the valid types for a given application--not in a static method of a singleton class that is global to all compilation units.  That was a feature of the Cohort MessageFactory that we should not lose--it's really the main benefit from the factory method pattern here:  the ability to supply better/different factory method implementations, not just remove a large switch statement.  This is preferable completely independent of how a given MessageFactory represents the message set.

**Comment by gregsfortytwo** [comment]
Yeah, IIRC we've actually inflated several things in order to accommodate our global messages (lots of things needing to include mdstypes, maybe, because they go into the MDS<->Client communications?). So we want to shrink it down.

**Comment by liewegas** [comment]
Closing this pull request since it appears to be abandoned.  If you it is not abandoned, and you are still interested in championing this change, please address any review comments so far and reopen the pull request.

**Pull request closed without merging**